lease.go

```diff
@@ -23,109 +23,119 @@ type WgRecord struct {
+		cidr:       cfg.WireguardIPNetwork,
+		deviceName: cfg.DeviceName,
+		filename:   cfg.LeasesFilename,
+		ip:         cfg.WireguardIPAddress,
 	}
 
-	lm := &FileLeaseManager{
-		wgRecords: wgRecords,
-		filename:  filename,
-		ip:        ip,
-		cidr:      cidr,
+	if err := lm.loadWgRecords(); err != nil {
+		return nil, err
 	}
 
-	if err := updateWgPeers(lm); err != nil {
+	if err := lm.updateWgPeers(); err != nil {
 		return nil, err
 	}
 
 	log.Println("Init complete")
 	return lm, nil
 }
 
-func loadWgRecords(r io.Reader) (map[string]*WgRecord, error) {
-	sc := bufio.NewScanner(r)
-	records := make(map[string]*WgRecord)
+func (lm *FileLeaseManager) loadWgRecords() error {
+	lm.wgRecordsMutex.Lock()
+	defer lm.wgRecordsMutex.Unlock()
+
+	lm.wgRecords = make(map[string]WgRecord)
+
+	r, err := os.Open(lm.filename)
+	if err != nil {
+		log.Printf("unable to open leases file: %v", err)
+		return nil
+	}
+	defer r.Close()
 
+	sc := bufio.NewScanner(r)
 	for sc.Scan() {
 		line := sc.Text()
 		if len(line) == 0 {
 			continue
 		}
 		tokens := strings.Fields(line)
 		if len(tokens) != 4 {
-			return nil, fmt.Errorf("malformed line, want 3 fields, got %d: %s", len(tokens), line)
+			return fmt.Errorf("malformed line, want 3 fields, got %d: %s", len(tokens), line)
 		}
 
 		username := tokens[0]
 		pubKey := tokens[1]
 		ipaddr := net.ParseIP(tokens[2])
 		// TODO: support v6?
 		if ipaddr.To4() == nil {
-			return nil, fmt.Errorf("expected an IPv4 address, got: %v", ipaddr)
+			return fmt.Errorf("expected an IPv4 address, got: %v", ipaddr)
 		}
 		expires, err := time.Parse(time.RFC3339, tokens[3])
 		if err != nil {
-			return nil, fmt.Errorf("expected time of exipry in RFC3339 format, got: %v", tokens[2])
+			return fmt.Errorf("expected time of exipry in RFC3339 format, got: %v", tokens[2])
 		}
 		if expires.After(time.Now()) {
-			records[username] = &WgRecord{
+			lm.wgRecords[username] = WgRecord{
 				PubKey:  pubKey,
 				IP:      ipaddr,
 				expires: expires,
 			}
 		}
 	}
-	return records, nil
+
+	log.Println("records loaded")
+	return nil
 }
 
-func (lm *FileLeaseManager) saveWgRecord(username string, record *WgRecord) error {
-	f, err := os.OpenFile(lm.filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+func (lm *FileLeaseManager) saveWgRecords() error {
+	lm.wgRecordsMutex.Lock()
+	defer lm.wgRecordsMutex.Unlock()
+	f, err := os.OpenFile(lm.filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
 	if err != nil {
 		return err
 	}
 	defer f.Close()
-	_, err = f.WriteString(username + " " + record.PubKey + " " + record.IP.String() + " " + record.expires.Format(time.RFC3339) + "\n")
-	if err != nil {
-		return err
-	}
-	err = f.Sync()
-	if err != nil {
-		return err
+	for username, record := range lm.wgRecords {
+		if _, err := fmt.Fprintf(f, "%s %s\n", username, record); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
 func (lm *FileLeaseManager) findNextAvailableIPAddress() (*net.IPNet, error) {
+	lm.wgRecordsMutex.Lock()
```

Comment:
I am not sure this is useful here, as this function doesn't update `lm.wgRecords` from which we construct the `allocatedIPs` list. Shouldn't the lock be at some place higher like `createOrUpdatePeer` function?

Reply:
You're accessing the map below so you really do need to lock it. However, you're right that the way locks are done right now in this code path are not sufficient. You can still have to requests receiving the same address I believe and the easiest fix for this would be to merge `findNextAvailableIPAddress` and `createOrUpdatePeer` and use a single lock for all of it.